File: 111. Minimum Depth of Binary Tree/solution_dfs_iterative.py
from typing import *

class TreeNode:
    """Definition for a binary tree node."""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        
        if not root:
            return 0

        stack = [(root, 1)]
        minimumDepth = float('inf')
        
        while stack:
            node, depth = stack.pop()

            if node:
                if not node.left and not node.right:
                    minimumDepth = min(minimumDepth, depth)
                    continue

                stack.append((node.left, depth + 1))
                stack.append((node.right, depth + 1))
        
        return minimumDepth

File: 111. Minimum Depth of Binary Tree/solution_dfs_iterative.js
/** Definition for a binary tree node. */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
    if (!root) return 0;
    
    const stack = [[root, 1]]; // [node, depth]
    let minimumDepth = Infinity;

    while (stack.length > 0) {
        const [node, depth] = stack.pop();

        if (node) {
            if (!node.left && !node.right) {
                minimumDepth = Math.min(minimumDepth, depth);
                continue;
            }

            stack.push([node.left, depth + 1]);
            stack.push([node.right, depth + 1]);
        }
    }

    return minimumDepth;
};